### PR TITLE
Allow customfields to be specified by their english names

### PR DESCRIPTION
**Pull request by nhemingway**
Changes expanded to cover both issue creation and retrieval

### CODE REVIEW

File: lib/JIRA/Client/Automated.pm [hunk 1]
```diff
@@ -252,9 +252,15 @@ sub _handle_error_response {
     my ($self, $response, $request) = @_;
 
     my $msg = $response->status_line;
-    $msg .= sprintf " %s", $response->decoded_content
+    #$msg .= sprintf " %s", $response->decoded_content
+    #    if $response->decoded_content;
+    use Data::Dump qw(pp);
```
Comment (frimicc) [inline lib/JIRA/Client/Automated.pm]: Can you remove debugging code from this and delete any commented out code that's been replaced?

**Comment by frimicc** [comment]
Wow, awesome patch. I never had the patience to mock the module to do disconnected testing. 

It looks to me like it is reading the custom field names once per issue and caching them, is that correct? Would it be worth just grabbing all the issues' custom fields at once, since they're all in the same JSON response from the server? 

How does this extra work (several extra API calls) affect the module's performance? I assume it's not bad, but it might be worth putting in a pass-through for any fields that have already been translated or a way to not lookup extra field names. My JIRA instance only has 2 custom fields and none with allowed values, so we just hard-code the mapping in our calling code. I suspect that might be the majority case.

Thanks!

**Comment by nhemingway** [comment]
Hi,

On 30 May 2016 at 19:30, Michael Friedman <notifications@github.com> wrote:

> Wow, awesome patch. I never had the patience to mock the module to do
> disconnected testing.
>

It is tedious, but I've found it enables testing a bunch of conditions that
can't be tested any other way.

> It looks to me like it is reading the custom field names once per issue
> and caching them, is that correct? Would it be worth just grabbing all the
> issues' custom fields at once, since they're all in the same JSON response
> from the server?
>
Not once per issue, but once per issuetype per project. Subsequent calls
for the same issuetype won't re-retrieve the custom field details.  I
decided that grabbing the custom field data for all issuetypes within a
project or worse, all issuetypes for all projects, would both balloon the
memory requirements, and also cause the introspection to take forever thus
slowing down the operation you're trying to achieve to a ridiculous degree.

> How does this extra work (several extra API calls) affect the module's
> performance? I assume it's not bad, but it might be worth putting in a
> pass-through for any fields that have already been translated or a way to
> not lookup extra field names. My JIRA instance only has 2 custom fields and
> none with allowed values, so we just hard-code the mapping in our calling
> code. I suspect that might be the majority case.
>
I agree that your case probably does match the majority case.  But is that
just because JIRA makes the alternative a PITA, so noone else has been
bloodyminded enough to do it :) ?

I'm using the patched modules to create a ticket and several sub-tickets,
and creating the tickets is far from the slowest part of the job.  But then
we've always complained that our JIRA instance is glacial at best, so any
additional slowdown wouldn't be noticed.

Not quite sure what you mean by "pass-through for any fields that have
already been translated"
* are you suggesting that I should simply bypass translating fields names
customfield_blah?  Whilst the patch doesn't do that, the time saved by
adding a bypass for such a field would be minimal because most of the time
will have been spent getting and parsing the metadata anyway.
* One alternative might be to specify a control field to the constructor,
to disable this auto-translate feature and not bother retrieving and
parsing the metadata at all.
* Or I could add a list of built-in field names for which no translation is
required.  This could then only retrieve the metadata should there be a
field that is neither a built-in nor an already translated field.  At that
point, users who have already translated their fields manually wouldn't pay
the price of the metadata retrieval.  The downside will be that the module
will then have a hard-coded list of fields that may become out-of-date
against different versions of JIRA.

Which of these (or something else) do you mean?

Neil

**Comment by nhemingway** [comment]
ping

**Comment by frimicc** [comment]
Sorry for the delay, summer is vacation time. 
I'm merging this now. How would you like to be credited? What email address should I include in the Changes file? Are you hemingway@cpan.org?

**Comment by nhemingway** [comment]
Sorry, didn't see your email.  Thanks for the credit anyway.  Any chance of getting a CPAN release out?

**Comment by frimicc** [comment]
The only thing holding up a release to CPAN is you telling me what email address you would like in the changes file. I always give credit where credit is due. What email would you like listed?

**Comment by nhemingway** [comment]
Ah, ok.  the cpan address will be fine thanks.  Sorry, thought you'd given
up waiting for me and merged to github anyway.

Neil

**Pull request merged**